mne/beamformer/_lcmv.py

```diff
@@ -102,52 +133,90 @@ def _apply_lcmv(data, info, tmin, forward, noise_cov, data_cov, reg,
+                                                   np.dot(Wk, Gk)))
+            idx_max = eig_vals.argmax()
+            max_ori = eig_vecs[:, idx_max]
             Wk[:] = np.dot(max_ori, Wk)
-            Ck = np.dot(max_ori, np.dot(Ck, max_ori))
+            Gk = np.dot(Gk, max_ori)
+
+            if weight_norm == 'nai':
+                # compute spatial filter for NAI
+                tmp = np.dot(Gk.T, np.dot(Cm_inv_sq, Gk))
+                denom = np.sqrt(noise * tmp)
+                Wk /= denom
+            elif weight_norm == 'unit-noise-gain':
+                # compute spatial filter for unit-noise gain
+                tmp = np.dot(Gk.T, np.dot(Cm_inv_sq, Gk))
+                denom = np.sqrt(tmp)
+                Wk /= denom
+
             is_free_ori = False
 
-        if is_free_ori:
-            # Free source orientation
-            Wk[:] = np.dot(linalg.pinv(Ck, 0.1), Wk)
+        # special case 2: NAI without orientation selection
+        elif weight_norm == 'nai' and pick_ori != 'max-power':
+            raise ValueError('Weight normalization with Neural activity index '
+                             'is not implemented yet with free or normal '
+                             'orientation.')
+
+        # all other combinations are handled here:
         else:
-            # Fixed source orientation
-            Wk /= Ck
+            if pick_ori == 'max-power':
+                eig_vals, eig_vecs = linalg.eigh(Ck)
+
+                # Choosing the eigenvector associated with the middle
+                # eigenvalue. The middle and not the minimal eigenvalue is used
+                # because MEG is insensitive to one (radial) of the three
+                # dipole orientations and therefore the smallest eigenvalue
+                # reflects mostly noise.
+                for ii in range(3):
+                    if ii != eig_vals.argmax() and ii != eig_vals.argmin():
+                        idx_middle = ii
+
+                # TODO: The eigenvector associated with the smallest eigenvalue
+                # should probably be used when using combined EEG and MEG data
+                max_ori = eig_vecs[:, idx_middle]
+
+                Wk[:] = np.dot(max_ori, Wk)
+                Ck = np.dot(max_ori, np.dot(Ck, max_ori))
+                is_free_ori = False
+
+            if is_free_ori:
+                # Free source orientation
+                Wk[:] = np.dot(linalg.pinv(Ck, 0.1), Wk)
+            else:
+                # Fixed source orientation
+                Wk /= Ck
 
     # Pick source orientation maximizing output source power
     if pick_ori == 'max-power':
         W = W[0::3]
 
-    # Preparing noise normalization
-    noise_norm = np.sum(W ** 2, axis=1)
-    if is_free_ori:
-        noise_norm = np.sum(np.reshape(noise_norm, (-1, 3)), axis=1)
-    noise_norm = np.sqrt(noise_norm)
+    if weight_norm == 'unit-noise-gain' and pick_ori != 'max-power':
```

Comment:
Yes, totally, all these if clauses are confusing. 
The normalization was only done for the max-power case so far. I think the free orientation case (the already implemented "noise normalization") here might need another formula check - whether it really is the vector beamformer unit-noise-gain solution. I checked with Sekihara & Nagarajan 2008 again and the formula here rather seems to converge with the scalar beamformer case (also see Ft-bugzilla link above). 
It is not implemented in Fieldtrip yet, so it's not possible to externally validate it. 
Can you help us (or at least me ;-) ) out here, @sarangnemo ?